Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs

```diff
@@ -1,237 +1,143 @@
-        TransformComponent? transformOrig = null;
-        FluidSpreaderComponent? spreader = null;
-
-        if (!Resolve(suid, ref puddleComponent, ref metadataOriginal, ref transformOrig, ref spreader, false))
-            return;
-
-        var prototypeName = metadataOriginal.EntityPrototype!.ID;
-        var visitedTiles = new HashSet<Vector2i>();
-
-        if (!_mapManager.TryGetGrid(transformOrig.GridUid, out var mapGrid))
-            return;
-
-        // skip origin puddle
-        var nextToExpand = new List<PuddlePlacer>(9);
-        ExpandPuddle(suid, visitedTiles, mapGrid, nextToExpand);
-
-        while (nextToExpand.Count > 0
-               && spreader.OverflownSolution.CurrentVolume > FixedPoint2.Zero)
+            Direction.North,
+            Direction.East,
+            Direction.South,
+            Direction.West,
+        };
+        foreach (var fluidMapData in EntityQuery<FluidMapDataComponent>())
         {
-            // we need to clamp to prevent spreading 0u fluids, while never going over spill limit
-            var divided = FixedPoint2.Clamp(spreader.OverflownSolution.CurrentVolume / nextToExpand.Count,
-                FixedPoint2.Epsilon, puddleComponent.OverflowVolume);
+            if (fluidMapData.Puddles.Count == 0 || _gameTiming.CurTime <= fluidMapData.GoalTime)
+                continue;
 
-            foreach (var posAndUid in nextToExpand)
+            var newIteration = new HashSet<EntityUid>();
+            foreach (var puddleUid in fluidMapData.Puddles)
             {
-                var puddleUid = GetOrCreate(posAndUid.Uid, prototypeName, mapGrid, posAndUid.Pos);
-
-                if (!TryComp(puddleUid, out PuddleComponent? puddle))
+                PuddleComponent? puddle = null;
+                MetaDataComponent? metaData = null;
+                TransformComponent? transform = null;
+                if (!Resolve(puddleUid, ref puddle, ref metaData, ref transform, false)
+                    || !_mapManager.TryGetGrid(transform.GridUid, out var mapGrid))
                     continue;
 
-                posAndUid.Uid = puddleUid;
-
-                if (puddle.CurrentVolume >= puddle.OverflowVolume) continue;
-
-                // -puddle.OverflowLeft is guaranteed to be >= 0
-                // iff puddle.CurrentVolume >= puddle.OverflowVolume
-                var split = FixedPoint2.Min(divided, -puddle.OverflowLeft);
-                _puddleSystem.TryAddSolution(
-                    puddle.Owner,
-                    spreader.OverflownSolution.SplitSolution(split),
-                    false, false, puddle);
-
-                // if solution is spent do not explore
-                if (spreader.OverflownSolution.CurrentVolume <= FixedPoint2.Zero)
-                    return;
+                var pos = transform.Coordinates;
+                var prototypeName = metaData.EntityPrototype!.ID;
+                var puddles = new List<PuddleComponent>(4);
```

Comment:
Arrays, ugh maybe, it would look a lot more uglier. 
Outside of loop would be pointless. I want to get up to 4 next  puddle and equalize them.

Reply:
Also, this begs the question since HashSet of Overflowing puddles is also used in a similar manner, should I clear it and add `newIteration` rather than creating and reassigning a new one?

Reply:
Pretty sure creating the array is going to be a lot slower than clearing it but I haven't checked.

Reply:
Done. HashSet is cleared then `UnionWith`>